tests/csv_datums/test_csv_sources.py

```diff
@@ -90,35 +124,198 @@ def source_interval():
 
 
 @pytest.fixture
-def source(source_interval):
-    return KrakenSource(pair="xbtusd", interval=source_interval)
+def source(source_interval, tmp_path):
+    return KrakenSource(trades_storage=tmp_path, pair="xbtusd", interval=source_interval, max_results=10)
 
 
 @pytest.fixture(autouse=True)
 def validation(monkeypatch):
-    monkeypatch.setattr(module_under_test, 'validate', lambda d: Validated(d))
+    class _Spy:
+        def __init__(self):
+            self.data = None
+            self.raises = None
+
+        def set_raises(self, raises):
+            self.raises = raises
+
+        def __call__(self, d, *args, **kwargs):
+            if self.raises:
+                raise self.raises("mocked data error")
+            self.data = d
+            return self.data
+
+    s = _Spy()
+    monkeypatch.setattr(module_under_test, 'validate', s)
+    return s
 
 
 @pytest.fixture(autouse=True)
 def adapter(monkeypatch):
     monkeypatch.setattr(module_under_test, 'KrakenAdapter', AdapterStub)
 
 
-def test_kraken_source_builds_correct_url(source, requests):
-    source.query(since=1559347200)
-    assert requests.received_get == make_get("https://api.kraken.com/0/public/Trades", pair="xbtusd",
-                                             since=1559347200000000000)
+@pytest.fixture(autouse=True)
+def local_time(monkeypatch, source_interval):
+    s = LocalTimeSpy()
+    monkeypatch.setattr(module_under_test, 'time', s)
+    return s
 
 
-def make_get(url, **params):
-    return GetRequest(url, params)
+@pytest.fixture
+def server_time(monkeypatch, source_interval):
+    s = ServerTimeStub()
+    s.set_current_time(1559347200 + source_interval + 1)
+    monkeypatch.setattr(module_under_test, 'KrakenServerTime', lambda: s)
+    return s
+
+
+@pytest.fixture
+def make_json(server_time, source_interval):
+    def factory(results, last=None, with_error=None):
+        res = make_default_json_response()
+        res['result'] = results
+        res['error'] = with_error or []
+        last = last or to_nano_sec(server_time.current_time + source_interval + 1)
+        res['result']['last'] = str(last)
+        return res
+
+    return factory
+
+
+@pytest.fixture
+def randomness(monkeypatch):
+    s = RandomnessSpy()
+    monkeypatch.setattr(module_under_test, 'random', s)
+    return s
 
 
-def test_kraken_source_returns_validated_and_adapted_data(source, source_interval, requests):
-    requests.set_get_response(json={'some': 'data'})
-    adapted = Adapted(json={'some': 'data'}, with_interval=source_interval)
-    assert source.query(since=1559347200) == Validated(csv_datums_from(adapted))
+def make_get(url, **params):
+    return GetRequest(url, params)
 
 
 def csv_datums_from(adapted):
     return CsvDatums(adapted.interval, adapted)
+
+
+def expand_to_pair(*values):
```

Comment:
`expand_to_trades`
maybe also add a from to or range option to specify the timestamps